### PR TITLE
KEP-4222: Support CBOR for generated clients based on client-go feature gating.

### PR DESCRIPTION
**Pull request by benluddy**
#### What type of PR is this?

/kind cleanup
/sig api-machinery

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:

This is still gated by the test-only client feature gate instance, which is not wired for runtime enablement via command-line flags or environment variables.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://kep.k8s.io/4222
```

### CODE REVIEW

File: staging/src/k8s.io/client-go/rest/config.go [hunk 1]
```diff
@@ -672,3 +675,34 @@ func CopyConfig(config *Config) *Config {
...
+	// TODO: Thoughts about this:
+
+	// This depends on a client-go feature gate, so it can't live in apimachinery. It's a
+	// substitute for conventional package-scoped variables (i.e. `var Codecs =
+	// NewCodecFactory(Scheme)`) when the caller doesn't really want to be responsible for codec
+	// manufacturing but they need a NegotiatedSerializer for their client's content config. Is
+	// there a better home for it?
+
+	// Is it worth caching these? Is there a realistic use case for creating clients so rapidly
+	// that the cost of NewCodecFactory becomes significant?
```
Comment (jpbetz) [inline staging/src/k8s.io/client-go/rest/config.go:689]: Maybe? https://pkg.go.dev/golang.org/x/sync/singleflight might work here?  Its not obvious to me just eyeballing this if it's needed or helpful.

File: staging/src/k8s.io/client-go/rest/config.go [hunk 2]
```diff
@@ -672,3 +675,34 @@ func CopyConfig(config *Config) *Config {
 	}
 	return c
 }
+
+func DefaultCodecFactoryForScheme(scheme *runtime.Scheme) serializer.CodecFactory {
```
Comment (jpbetz) [inline staging/src/k8s.io/client-go/rest/config.go:679]: The approach in this PR looks reasonable to me.  I haven't thought carefully about if this is the best package for this function, but it doesn't seem unreasonable.

File: staging/src/k8s.io/code-generator/cmd/client-gen/generators/generator_for_group.go [hunk 3]
```diff
@@ -253,7 +253,7 @@ func setConfigDefaults(config *$.restConfig|raw$) error {
 	gv := $.SchemeGroupVersion|raw$
 	config.GroupVersion =  &gv
 	config.APIPath = $.apiPath$
-	config.NegotiatedSerializer = $.CodecsWithoutConversion|raw$()
+	config.NegotiatedSerializer = $.restDefaultCodecFactoryForScheme|raw$($.Scheme|raw$).WithoutConversion()
```
Comment (jpbetz) [inline staging/src/k8s.io/code-generator/cmd/client-gen/generators/generator_for_group.go:256]: All the generator changes look fine.

File: staging/src/k8s.io/client-go/rest/config.go (hunk 1 above)
Comment (deads2k) [reply staging/src/k8s.io/client-go/rest/config.go:689]: > Maybe? https://pkg.go.dev/golang.org/x/sync/singleflight might work here? Its not obvious to me just eyeballing this if it's needed or helpful.

Perhaps a refinement before going beta.  @benluddy I suggest opening a beta requirement PR to the KEP so we dont' lose it. New clients are created for some use-cases around "now connect with these credentials".

File: staging/src/k8s.io/client-go/rest/config.go [hunk 4]
```diff
@@ -672,3 +675,34 @@ func CopyConfig(config *Config) *Config {
...
+	// This depends on a client-go feature gate, so it can't live in apimachinery. It's a
+	// substitute for conventional package-scoped variables (i.e. `var Codecs =
+	// NewCodecFactory(Scheme)`) when the caller doesn't really want to be responsible for codec
+	// manufacturing but they need a NegotiatedSerializer for their client's content config. Is
+	// there a better home for it?
+
+	// Is it worth caching these? Is there a realistic use case for creating clients so rapidly
+	// that the cost of NewCodecFactory becomes significant?
+
+	if features.TestOnlyFeatureGates.Enabled(features.TestOnlyClientAllowsCBOR) {
```
Comment (deads2k) [inline staging/src/k8s.io/client-go/rest/config.go:691]: Can we invert this check and simply return `scheme.Codecs` when the check is false?

File: staging/src/k8s.io/client-go/rest/client.go [hunk 5]
```diff
@@ -115,14 +119,54 @@ func NewRESTClient(baseURL *url.URL, versionedAPIPath string, config ClientConte
...
+	}
+
+	if mediatype, _, err := mime.ParseMediaType(content.ContentType); err == nil && mediatype == "application/cbor" {
+		content.ContentType = "application/json"
+	}
+
+	clauses := goautoneg.ParseAccept(content.AcceptContentTypes)
+	scrubbed := false
+	for i, clause := range clauses {
+		if clause.Type != "application" || clause.SubType != "cbor" {
```
Comment (deads2k) [inline staging/src/k8s.io/client-go/rest/client.go:142]: nit: demorgans please

Comment (benluddy) [reply staging/src/k8s.io/client-go/rest/client.go]: done

File: test/integration/client/client_test.go [hunk 6]
```diff
@@ -1347,3 +1359,494 @@ func TestExtractModifyApply_ForceOwnership(t *testing.T) {
...
+			name:                    "cbor allowed and not preferred client forces protobuf",
+			served:                  true,
+			allowed:                 true,
+			preferred:               false,
+			wantRequestContentType:  "application/vnd.kubernetes.protobuf",
+			wantRequestAccept:       "application/vnd.kubernetes.protobuf,application/json",
+			wantResponseContentType: "application/vnd.kubernetes.protobuf",
+			wantResponseStatus:      http.StatusCreated,
+			wantStatusError:         false,
+			doRequest:               DoRequestWithProtobufPreferredGeneratedClient,
```
Comment (deads2k) [inline test/integration/client/client_test.go:1441]: why does a test labelled "not preferred client forces protobuf" use a client called `DoRequestWithProtobufPreferredGeneratedClient`?

Comment (benluddy) [reply test/integration/client/client_test.go:1441]: The "client prefers CBOR" gate that controls the default request encoding is disabled.

File: test/integration/client/client_test.go [hunk 7]
```diff
@@ -1347,3 +1359,494 @@ func TestExtractModifyApply_ForceOwnership(t *testing.T) {
...
+			preferred:               true,
+			wantRequestContentType:  "application/cbor",
+			wantRequestAccept:       "application/cbor, */*",
+			wantResponseContentType: "application/cbor",
+			wantResponseStatus:      http.StatusCreated,
+			wantStatusError:         false,
+			doRequest:               DoRequestWithGeneratedClient,
+		},
+		{
+			name:                    "send cbor accept both get 415",
```
Comment (deads2k) [inline test/integration/client/client_test.go:1516]: I had forgotten about this one.  good.

File: test/integration/client/client_test.go [hunk 8]
```diff
@@ -1347,3 +1359,494 @@ func TestExtractModifyApply_ForceOwnership(t *testing.T) {
...
+			configuredAccept:        "application/cbor",
+			wantRequestContentType:  "application/cbor",
+			wantRequestAccept:       "application/cbor",
+			wantResponseContentType: "application/cbor",
+			wantResponseStatus:      http.StatusCreated,
+			wantStatusError:         false,
+			doRequest:               DoRequestWithGeneratedClient,
+		},
+		{
+			name:                    "force disable actively configured cbor",
```
Comment (deads2k) [inline test/integration/client/client_test.go:1554]: very good

File: test/integration/client/client_test.go (hunk 6 above)
Comment (deads2k) [reply test/integration/client/client_test.go:1441]: Thanks.

**Review by dgrisonnet — APPROVED** [review]
/approve

**Pull request merged**